Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,204 @@
+ * cancellation via the cancel method.
+ */
+class DelayedOperation {
+ public:
+  /**
+   * Cancels the operation if it hasn't already been executed or canceled.
+   *
+   * As long as the operation has not yet been run, calling `Cancel()` (from an
+   * operation already running on the dispatch queue) provides a guarantee that
+   * the operation will not be run.
+   */
+  void Cancel();
+
+ private:
+  // Don't allow callers to create their own `DelayedOperation`s.
+  friend class AsyncQueue;
+  explicit DelayedOperation(
+      const std::shared_ptr<detail::DelayedOperationImpl>& instance)
+      : handle_{instance} {
+  }
+
+  std::weak_ptr<detail::DelayedOperationImpl> handle_;
+};
+
+class AsyncQueue {
+ public:
+  using Milliseconds = std::chrono::milliseconds;
+  using Operation = std::function<void()>;
+
+  explicit AsyncQueue(const dispatch_queue_t dispatch_queue)
+      : dispatch_queue_{dispatch_queue} {
+  }
+
+  /**
+   * Asserts that we are already running on this queue (actually, we can only
+   * verify that the queue's label is the same, but hopefully that's good
+   * enough).
+   */
+  void VerifyIsCurrentQueue() const;
+
+  /**
+   * Declares that we are already executing on the correct `dispatch_queue_t`
+   * and would like to officially execute code on behalf of this `AsyncQueue`.
+   * To be used only when called back by some other API directly onto our queue.
+   * This allows us to safely dispatch directly onto the worker queue without
+   * destroying the invariants this class helps us maintain.
+   */
+  void EnterCheckedOperation(const Operation& operation);
+
+  /**
+   * Same as `dispatch_async()` except it asserts that we're not already on the
+   * queue, since this generally indicates a bug (and can lead to re-ordering of
+   * operations, etc).
+   *
+   * @param operation The operation to run.
+   */
+  void Enqueue(const Operation& operation);
+
+  /**
+   * Unlike `Enqueue`, this method does not require you to dispatch to a
+   * different queue than the current one (thus it is equivalent to a raw
+   * `dispatch_async()`).
+   *
+   * This is useful, e.g. for dispatching to the user's queue directly from user
+   * API call (in which case we don't know if we're already on the user's queue
+   * or not).
+   *
+   * @param operation The operation to run.
+   */
+  void EnqueueAllowingSameQueue(const Operation& operation);
+
+  /**
+   * Schedules an operation after the specified delay.
+   *
+   * Unlike `Enqueue`, this method does not require you to dispatch to a
+   * different queue than the current one.
+   *
+   * The returned `DelayedOperation` handle can be used to cancel the operation
+   * prior to its running.
+   *
+   * @param operation The operation to run.
+   * @param delay The delay after which to run the operation.
+   * @param timer_id A `TimerId` that can be used from tests to check for the
+   *     presence of this operation or to schedule it to run early.
```

Comment:
Rephrased the comment, PTAL.

> nit: I'm not sure that's technically correct since the implementation of EnqueueAfterDelay will assert if there's two operations with the same timer_id

Right.

> Curiously, only obj-c has this parameter. Android and typescript do not

They both have it, e.g. [TypeScript](https://github.com/firebase/firebase-js-sdk/blob/0fa319e5e019dd0d40ab441d2ff9f8f6d4724e43/packages/firestore/src/util/async_queue.ts#L237).

Reply:
> Rephrased the comment, PTAL.

I don't see it? (Last commit I see is from yesterday.)

Oh, my mistake. My client was out of date. By a lot. :)

Reply:
Sorry, it's actually my mistake. Pushed the commit now.

Reply:
I see the comment now. lgtm.